src/Microsoft.Performance.SDK.Runtime/DTO/TableConfigurationsSerializer.cs

```diff
@@ -94,15 +96,32 @@ public class TableConfigurationsSerializer
 
                 var desrailizedObject = serializer.ReadObject(stream);
 
-                if (desrailizedObject is ISupportUpgrade<PrebuiltConfigurations> supportUpgrade)
+                if (desrailizedObject is not PrebuiltConfigurationsBase configurationsBase)
                 {
-                    prebuiltConfigurations = supportUpgrade.Upgrade();
+                    return Enumerable.Empty<Processing.TableConfigurations>();
                 }
-                else if (desrailizedObject is PrebuiltConfigurations latestConfigs)
+
+                // Ensure we don't get stuck in a loop by tracking visited configurations
+                HashSet<PrebuiltConfigurationsBase> visited = new HashSet<PrebuiltConfigurationsBase>();
```

Comment:
When will there be a cycle? If a configuration wants to upgrade to itself or downgrade then it shouldn't implement `ISupportUpgrade`?

Reply:
It would happen if someone implements the upgrading wrong. I.e. you upgrade from 1.0 to 2.0 then 2.0 "upgrades" to 1.0.

This is mostly to prevent an infinite loop below

Reply:
So sounds like this isn't actually possible with our existing code? Usually it's better to guard potential developer mistakes with tests than adding runtime guard/checks. For this case we can have a parameterized (by the configuration type) test that verify they won't create infinite upgrade cycle (by using timeout)

Reply:
We have tests that perform a round-trip serialize/deserialize from the pre-v1.0 configuration to the latest, so we are already verifying we don't hit an infinite loop. I am fine removing this check given that we have the tests, but I usually prefer more defensive programming where I assume tests like that **don't** exist. I would prefer a future bug where tables don't get deserialized correctly then a bug where an application hangs indefinitely

Reply:
Although, if we are fine keeping this, I should probably update this to be smarter than a standard reference equality check. I think the current implementation only works if two returned upgraded versions are the same instance

Reply:
I would say we remove this. It helps reduce the complexity of the production code given we have the tests

Reply:
We can leave a comment if you feel it's something that future developers should pay attention to

Reply:
Changed this to check the version returned by `UpgradeToNext`